src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,304 @@
+            if (preferredWriteAsyncMethod == null)
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                IOperation parentOperation = invocation.Parent;
+                if (parentOperation == null)
+                {
+                    return;
+                }
+                // Only accept these two cases:
+                // - await {WriteAsync()|ReadAsync()}
+                // - await {WriteAsync()|ReadAsync()}.ConfigureAwait()
+                if (parentOperation.Kind == OperationKind.Await ||
+                        (parentOperation.Parent != null && parentOperation.Parent.Kind == OperationKind.Await &&
+                         parentOperation.Parent is IAwaitOperation awaitOperation && awaitOperation.Operation is IInvocationOperation grandparentInvocation &&
+                            (grandparentInvocation.TargetMethod.OriginalDefinition.Equals(configureAwaitMethod) ||
+                             grandparentInvocation.TargetMethod.OriginalDefinition.Equals(genericConfigureAwaitMethod))))
+                {
+                    // Verify if the current method's type is or inherits from Stream
+                    if (!IsDefinedBy(invocation.TargetMethod, streamType, out IMethodSymbol method))
+                    {
+                        return;
+                    }
+
+                    DiagnosticDescriptor rule;
+                    string ruleMessageMethod;
+                    string ruleMessagePreferredMethod;
+
+                    // Verify if the method is an undesired Async overload
+                    if (method.Equals(undesiredReadAsyncMethod) || method.Equals(undesiredReadAsyncMethodWithCancellationToken))
+                    {
+                        rule = PreferStreamReadAsyncMemoryOverloadsRule;
+                        ruleMessageMethod = undesiredReadAsyncMethod.Name;
+                        ruleMessagePreferredMethod = preferredReadAsyncMethod.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+                    }
+                    else if (method.Equals(undesiredWriteAsyncMethod) || method.Equals(undesiredWriteAsyncMethodWithCancellationToken))
+                    {
+                        rule = PreferStreamWriteAsyncMemoryOverloadsRule;
+                        ruleMessageMethod = undesiredWriteAsyncMethod.Name;
+                        ruleMessagePreferredMethod = preferredWriteAsyncMethod.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+                    }
+                    else
+                    {
+                        return;
+                    }
+
+                    context.ReportDiagnostic(
+                        invocation.CreateDiagnostic(rule, ruleMessageMethod, ruleMessagePreferredMethod));
+                }
+            }, OperationKind.Invocation);
```

Comment:
If we are only flagging `await` operations, then we should change the RegisterOperationAction callback to register `OperationKind.Await` instead of `OperationKind.Invocation`

Reply:
@mavasani Doing this will force me to change my code to analyze an `IAwaitOperation` at the beginning of `RegisterOperationAction`, which does not expose a `TargetMethod` property.
Can we keep `OperationKind.Invocation`? The await is simply a scope restriction of the analyzer, not the actual target of the analyzer. The target is the method invocation itself.

Reply:
I'll try to find a way to retrieve the child invocation operation when detecting an await operation.

Reply:
Addressed. Please take a look.